Network specific conf sections

ajtowns opened the pull request:
The weekly meeting on [2017-12-07](http://www.erisian.com.au/meetbot/bitcoin-core-dev/2017/bitcoin-core-dev.2017-12-07-19.00.log.html) discussed allowing options to bitcoin to have some sensitivity to what network is in use. @theuni suggested having sections in the config file:

    <cfields> an alternative to that would be sections in a config file. and on the
              cmdline they'd look like namespaces. so, [testnet] port=5. or -testnet::port=5.

This approach is (more or less) supported by `boost::program_options::detail::config_file_iterator` -- when it sees a `[testnet]` section with `port=5`, it will treat that the same as "testnet.port=5". So `[testnet] port=5` (or `testnet.port=5` without the section header) in bitcoin.conf and `-testnet.port=5` on the command line.

The other aspect to this question is possibly limiting some options so that there is no possibility of accidental cross-contamination across networks. For example, if you're using a particular wallet.dat on mainnet, you may not want to accidentally use the same wallet on testnet and risk reusing keys.

I've set this up so that the `-addnode` and `-wallet` options are `NETWORK_ONLY`, so that if you have a bitcoin.conf:

    wallet=/secret/wallet.dat
    upnp=1

and you run `bitcoind -testnet` or `bitcoind -regtest`, then the `wallet=` setting will be ignored, and should behave as if your bitcoin.conf had specified:

    upnp=1

    [main]
    wallet=/secret/wallet.dat

For any `NETWORK_ONLY` options, if you're using `-testnet` or `-regtest`, you'll have to add the prefix to any command line options. This was necessary for `multiwallet.py` for instance.

I've left the "default" options as taking precedence over network specific ones, which might be backwards. So if you have:

    maxmempool=200
    [regtest]
    maxmempool=100

your maxmempool will still be 200 on regtest. The advantage of doing it this way is that if you have `[regtest] maxmempool=100` in bitcoin.conf, and then say `bitcoind -regtest -maxmempool=200`, the same result is probably in line with what you expect...

The other thing to note is that I'm using the chain names from `chainparamsbase.cpp` / `ChainNameFromCommandLine`, so the sections are `[main]`, `[test]` and `[regtest]`; not `[mainnet]` or `[testnet]` as might be expected.

Thoughts? Ping @MeshCollider @laanwj @jonasschnelli @morcos

MeshCollider commented:
Ah you beat me to it, will take a look as soon as I can

jnewbery commented:
Great. Thanks for tackling this @ajtowns. A few high-level suggestions:

- I agree with @MeshCollider that the order of precedence should be Command Line > Config File network-specific > Config File default.
- I think there shouldn't be network-specific command-line options (that's almost an implication of the above). I agree with MeshCollider that the user should call `bitcoind` with `-wallet`, not `-regtest.wallet` (side-note - choosing a wallet from the wrong network is perhaps less of a problem than you might assume, since `-wallet` is already kind of network-specific as it refers to the name of the wallet in the particular network directory).
- as well as `NETWORK_ONLY`, perhaps we want a `DEFAULT_ONLY` for options that shouldn't be overriden in the network-specific config. Namely those options would be `-regtest` and `-testnet`. For example don't want the following pathological config file to be valid:
```
testnet
[testnet]
regtest
[regtest]
notestnet
noregtest
```
- taking this further, perhaps we should deprecate the use of `testnet` and `regtest` in the config file entirely, so you can only select the network by using a command-line option?
- I've avoided nitting your code at this early point, but I'd point you at the developer notes here: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md, specifically symbol naming.  Your new `ArgsManager` members should be prefixed `m_` for example.

jonasschnelli commented:
Concept ACK.
Nice work!

Agree with @MeshCollider and @jnewbery (except the point of deprecating the use of testnet/regtest in config file).

ajtowns commented:
Okay, getting that behaviour requires marginally more invasive changes, but I think it's worth it (and they're not *that* invasive). New patch series accompanying this comment which:

 * has arg precedence being: command-line arguments, net-specific config file options, default config file options
 * will only use config file options for -addnode and -wallet if you're on mainnet or put them in the right [regtest] or [testnet] section or add a `regtest.` or `testnet.` prefix
 * only allows -testnet and -regtest to be specified in the default section or on the command line (ie, [regtest] testnet=1 silliness is silently ignored)
 * for single-value options, precedence is (a) last value specified on the command line, (b) first value specified in the network section of the config file, and finally (c) first value specified in the default section of the config file. the last vs first difference between cli and conf is weird, but matches current behaviour. the code would be slightly improved by making it consistent.

I think that should behave pretty much as expected -- specifying things on the command line will always take effect, having a testnet specific config and just invoking "bitcoind -conf=testnet.conf" should work fine, having separate configs for mainnet, regtest and testnet in a single file with different options should work too.

From the code side of things, I've made a bunch of changes:

 * I've dropped mapArgs, and split mapMultiArgs into mapOverrideArgs (for cli arguments are ForceSetArg) and mapConfigArgs (for config files)
 * I've listed the net-specific options explicitly in util.cpp rather than having each use of the options specify that they're net-specific; seems simpler and much more robust.
 * I've moved ChainNameFromCommandLine from chainparamsbase into util, so it can have more precise access to how -testnet/-regtest were specified
 * I've dropped the filename argument from ReadConfigFile and split most of the logic into ReadConfigStream, which makes it more unit-testable

I've added some unit tests, but they're a bit basic; I haven't done a functional test either. (And it's still a bit early for detailed code style nits, though I've tried to make a token effort to avoid them :) )

ajtowns commented:
Okay, here's something for first pass reviews I think.

The first five commits do some refactoring on ChainNameFromCommandLine (moving it from chainparamsbase.cpp into util.cpp:ArgsManager) and ReadConfigFile, primarily to make it easier to add tests for both of them.

    c5b5997e1 Move ChainNameFromCommandLine into ArgsManager
    eedd60951 [tests] Add unit tests for ChainNameFromCommandLine
    f1665f160 Separate out ReadConfigStream from ReadConfigFile
    0e78382c3 [tests] Add unit tests for ReadConfigStream
    d4367e1c6 [tests] Check ChainNameFromCommandLine works with config entries

The next commit switches from mapArgs and mapMultiArgs (for single and multi value options) to mapConfigArgs and mapOverrideArgs (for config file options and commandline/forceset options).

    bdb0c2857 ArgsManager: keep command line and config file arguments separate

The next commit is the first one that should change behaviour, and does most of the work. Unit test, and update to functional test are in the following commits.

    0e6679466 ArgManager: add support for config sections
    7a51198d3 [tests] Unit tests for config file sections
    359c78e2e [tests] Use regtest section in functional tests configs

The next commit changes the behaviour of the addnode, connect, port, bind, rpcport, rpcbind and wallet options: when they're specified in the config file they'll only apply to mainnet unless you use a [regtest] or [test] header (or a "regtest." or "test." prefix).

    7be1ddd31 ArgsManager: limit some options to only apply on mainnet when in default section
    c031709db [tests] Unit tests for section-specific config entries

Penultimately, there's special handling for the -regtest and -testnet options, so that ChainNameFromCommandLine will ignore bogus entries like "[regtest] testnet=1" or "[regtest] noregtest" and won't return weird results if invoked after the initial call to Select(Base)Params() with a pathological config file like that. cf jnewbery's comments in https://github.com/bitcoin/bitcoin/pull/11862#issuecomment-350753811

    2933f6016 ArgsManager: special handling for -regtest and -testnet
    22da57f19 [tests] Unit tests for -testnet/-regtest in [test]/[regtest] sections

And finally there's also an update for the release notes.

    ee91028e7 Add config changes to release notes

Notes:

I've called the member variables things like "m_mapConfigArgs" rather than fully snake-case map_config_args or similar as a compromise between the coding style and what's already in use in the class.

I needed a few helper functions with access to protected ArgsManager member variables; rather than add them to util.h directly, I made a friend class for them, ArgsManagerHelper. It's not crazy pretty but it seems alright. Maybe they should go in util_helper.h or similar though?

I got the locks wrong for cs_args and csPathCached initially -- was tempted to add "AssetLockNotHeld(gArgs.cs_args)" to GetDataDir to help avoid this, but it runs into two problems: in some threads GetDataDir() is called before any locks are acquired leaving lockstack==nullptr and causing a segfault (which can easily be avoided by just adding the assert after LOCK(csPathCached)); and cs_args is protected so it would've required another ArgsManagerHelper functions.

The datadir option is a little tricky. Currently you have three options:

  - don't specify -datadir anywhere, just use the default
  - specify -datadir on the command line
  - specify -datadir in the config file only (the config file may be located via the default datadir)

With these patches it would be possible to have a fourth option: specify datadir in the applicable network section of the configfile, allowing a single config file to specify different datadirs for regtest and testnet. At present the code locks datadir into place before selecting the network, so that hasn't been implemented in this PR.

The fact that -addnode, -connect, etc only apply to main net is specified in util.cpp rather than in the files that care about the options. This is something that can be cleaned up in the rework of argument handling that @MeshCollider has been working on.

I think the approach makes sense at this point, so reviews appreciated and nits welcome I guess!

laanwj commented:
Was already wondering why this was closed!

MeshCollider commented:
I think this is far enough along to have the brainstorming tag removed because at least to me it feels almost unready to be reviewed if it has that tag, more of a high-level discussion thing. 

Also could someone please assign me to review this so I remember :) @laanwj

jnewbery commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
+public:
+    inline static bool UseDefaultSection(const ArgsManager& am, const std::string& strArg)
+    {
+        if (am.m_strSection == CBaseChainParams::MAIN)
```
style nit: if clauses should either be on the same line as the if conditional, or with braces (https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md)

jnewbery reviewed (commented):
I've reviewed up to 940e5bc61f861aae97983d4be52ff87c101e29cc. A couple of comments inline.

jnewbery commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
+/** Internal helper functions for ArgsManager */
+class ArgsManagerHelper {
+public:
+    inline static bool UseDefaultSection(const ArgsManager& am, const std::string& strArg)
```
Perhaps add a comment for why we need this function (and another comment lower down next to `m_setSectionOnlyArgs` explaining why those particular arguments shouldn't be applied to non-mainnet configs by default).

jnewbery commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
+    }
+};
+
+ArgsManager::ArgsManager(void) :
```
Does it makes more sense to have this as a file-level constant rather than a member of ArgsManager?

ajtowns replied:
Having it be a member of ArgsManager makes it easier to deal with unit tests. But slightly longer term, they should be file-level constants in the files that use the arguments, rather than util.cpp; but that's future work.

jnewbery commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
+    /** Convert regular argument into the section-specific setting */
+    inline static std::string SectionArg(const ArgsManager& am, const std::string& strArg)
+    {
+        assert(strArg.length() > 1 && strArg[0] == '-');
```
This assert can be hit. Try running `feature_pruning.py` for example:

```
bitcoind: util.cpp:456: static std::__cxx11::string ArgsManagerHelper::SectionArg(const ArgsManager&, const string&): Assertion `strArg.length() > 1 && strArg[0] == '-'' failed.
```

ajtowns replied:
I believe this is catching a legitimate bug in the caller; confirming at the minute.

jnewbery commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
+                return true;
+        }
+
+        if (UseDefaultSection(am, strArg)) {
```
Is it possible to log when a default section config is not used? It could be confusing for users who specify `-addnode`, `-connect`, etc in the default section, expecting the config to be picked up.

ajtowns replied:
As in "Warning: you're running on testnet, so your -addnode setting didn't get applied" ?

laanwj replied:
Or "Warning: Setting X specified outside configuration section is only applied on mainnet"

jnewbery reviewed (commented):
A bunch of review comments inline. The new assert in `SectionArg()` can be hit, causing the node to crash.

I also think it would be a good idea to update the help text for `-addnode`, `-connect` to explain that they're not used for non-mainnet config unless explicitly specified.

Note that this can be a breaking api change for users who have `-addnode`, `-connect`, etc in their config files for testnet and regtest.

ajtowns commented on src/util.cpp:
```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
 }
 
+/** Internal helper functions for ArgsManager */
+class ArgsManagerHelper {
```
Yeah: It's a class so that it can be declared a friend of ArgsManager, which allows the functions in the class to access private/protected members of ArgsManager.

These functions could just be private member functions of ArgsManager, but that would mean bumping util.h every time any of them need to be changed, which causes a lot of unnecessary recompilation.

jnewbery commented:
> Let's ... try to get this in for 0.17.0

Current implementation is blocked on issue #12640 , which is fixed in #12756. Please review that (simple) PR if you want this to be merged :)

jnewbery commented:
#12756 is in master and I wanted to test this, so I rebased. There was a trivial merge conflict in feature_config_args.py. It might be quicker if you grab my rebased branch here: https://github.com/jnewbery/bitcoin/tree/pr11862.1

I can confirm with #12756, we no longer crash in any of the functional tests.

I think there are still a few review comments for you to address.

ajtowns commented:
Should address most of the comments, including the style nits.

I haven't updated the help for -addnode or -connect; not sure what to update it to... Also have not added a log message warning when no config option has been found despite a config option being present in the default section. Do have an idea how to do that, so may add it tomorrow.

Dropped [wip] from the subject; someone might want to drop the brainstorming label, 'cause I can't. :)

jnewbery commented:
> may add it tomorrow.

I'll review once you've decided how to tackle this :)

ajtowns commented:
Took the opportunity to bump the dates on the commits to encourage github to list them in the right order.

It now gives a warning at startup if you're on testnet or regtest, have one of the section-only arguments in the default section of the config file, and haven't overridden that argument either on the command line or in the appropriate config section. The warnings are checked via InitParameterInteraction() to avoid updating multiple places and it's kind-of a parameter interaction.

travis failures seem to just be the currently expected ones -- mingw32 tests timing out, and the one fixed by #12821.

ajtowns committed:
ArgsManager: keep command line and config file arguments separate

ajtowns committed:
ArgsManager: drop m_negated_args

When a -nofoo option is seen, instead of adding it to a separate
set of negated args, set the arg as being an empty vector of strings.

This changes the behaviour in some ways:
 - -nofoo=0 still sets foo=1 but no longer treats it as a negated arg
 - -nofoo=1 -foo=2 has GetArgs() return [2] rather than [2,0]
 - "foo=2 \n -nofoo=1" in a config file no longer returns [2,0], just [0]
 - GetArgs returns an empty vector for negated args

ajtowns committed:
ArgsManager: support config file sections

ajtowns committed:
[tests] Unit tests for config file sections

ajtowns committed:
[tests] Use regtest section in functional tests configs

ajtowns committed:
ArgsManager: limit some options to only apply on mainnet when in default section

When specified in bitcoin.conf without using the [regtest] or [test]
section header, or a "regtest." or "test." prefix, the "addnode",
"connect", "port", "bind", "rpcport", "rpcbind", and "wallet" settings
will only be applied when running on mainnet.

ajtowns committed:
ArgsManager: Warn when ignoring network-specific config setting

When network-specific options such as -addnode, -connect, etc are
specified in the default section of the config file, but that setting is
ignored due to testnet or regtest being in use, and it is not overridden
by either a command line option or a setting in the [regtest] or [test]
section of the config file, a warning is added to the log, eg:

  Warning: Config setting for -connect only applied on regtest network when in [regtest] section.

ajtowns committed:
[tests] Unit tests for network-specific config entries

ajtowns committed:
ArgsManager: special handling for -regtest and -testnet

ajtowns committed:
[tests] Unit tests for -testnet/-regtest in [test]/[regtest] sections

ajtowns committed:
Add config changes to release notes

ajtowns commented:
Rebased again. #12713 made this a bit more complicated -- how negated options (-nofoo) work with the config files is (in my opinion) a bit confusing as implemented, and coping with negated options and sections makes it a bit more confusing. As a consequence, I've broken out the first set of refactoring/testing patches into its on PR (#12878) which includes some additional tests for the current behaviour of negated options. For this PR, the way I've handled things is to make "-nofoo" empty the vector of values for "foo", and to consider an empty vector an indication that the arg is negated. I think that ends up behaving reasonably logically, and the changes are hopefully indicated in the corresponding unit tests.

jnewbery commented:
I've reviewed #12878. Will review this once that PR is merged and this is rebased.

jtimon commented:
Concept ACK. 
Right now, If I introduce a regtest2 section, for example, it will simply be ignored, correct?
I'm wondering how this will interact with https://github.com/bitcoin/bitcoin/pull/8994

ajtowns commented:
@jtimon If you put a `[regtest2]` section in bitcoind.conf today, it will be ignored -- under the hood the options are just prefixed with the section, so you'd just be setting `-regtest2.foo=1` which would never get looked up. I think as far as #8994 goes, the call to SelectConfigSection() that this PR adds inside SelectBaseParams() should keep the config section used (by this PR) in sync with the chain used (by your PR), so things should work out okay.

ajtowns commented:
Rebased post #12878 merge.

jnewbery commented on src/chainparamsbase.cpp:
```diff
@@ -48,4 +48,5 @@ std::unique_ptr<CBaseChainParams> CreateBaseChainParams(const std::string& chain
 void SelectBaseParams(const std::string& chain)
 {
     globalChainBaseParams = CreateBaseChainParams(chain);
+    gArgs.SelectConfigSection(chain);
```
I'm not sure if it makes sense to call this from `SelectBaseParams()`. This function is about setting the blockchain parameters, not the global node config.

Doesn't it make sense just to set `gArgs.m_section` directly at the end of `ReadConfigFile()`?

ajtowns replied:
I don't think so -- `ReadConfigFile()` only gets to know `-testnet` was set, you need to call `GetChainName()` to work out that means `chain=="test"`, and the place that's currently called is when passing the arguments to `SelectParams` or `SelectBaseParams`. Also, `SelectBaseParams` is already updating `globalChainBaseParams` which seems like "global node config" to me. :)

Changing that might not be a bad idea (and I think it might benefit jtimon's #8994), but I think it could wait for a different PR?

jnewbery replied:
fine. Can wait for a future PR.

jnewbery commented on src/util.cpp:
```diff
@@ -455,6 +455,99 @@ static bool InterpretBool(const std::string& strValue)
+
+    /** Determine whether to use config settings in the default section,
+     *  See also comments around ArgsManager::ArgsManager() below. */
+    inline static bool UseDefaultSection(const ArgsManager& am, const std::string& arg)
```
nanonit: use `inline static` or `static inline` consistently :)

jnewbery commented on src/util.h:
```diff
@@ -223,17 +224,35 @@ inline bool IsSwitchChar(char c)
-    std::unordered_set<std::string> m_negated_args;
+    std::map<std::string, std::vector<std::string>> m_override_args;
+    std::map<std::string, std::vector<std::string>> m_config_args;
+    std::string m_section;
```
I think this variable could be named better (and commented). If I'm not mistaken, this is the network that the node has been configured to run on (mainnet, testnet or regtest). Would a better name be `m_network`?

jnewbery commented on src/util.cpp:
```diff
@@ -464,30 +557,84 @@ static bool InterpretBool(const std::string& strValue)
+
+    for (const auto& arg : m_section_only_args) {
+        // if it's okay to use the default section for this chain, don't worry
+        if (ArgsManagerHelper::UseDefaultSection(*this, arg)) return;
```
This will only return true if `am.m_section == CBaseChainParams::MAIN`. Why don't you explictly check for that above the loop and return?

jnewbery commented on src/test/util_tests.cpp:
```diff
@@ -463,14 +566,26 @@ BOOST_AUTO_TEST_CASE(util_GetChainName)
-    test_args.ParseParameters(2, (char**)argv_regtest);
+    test_args.ParseParameters(2, (char**)argv_testnet);
     test_args.ReadConfigString(testnetconf);
-    BOOST_CHECK_THROW(test_args.GetChainName(), std::runtime_error);
```
You've removed this test that `GetChainName()` will throw if both `regtest` and `testnet` have been set. Is that intentional?

ajtowns replied:
No, I think that was a mistake while rebasing.

jnewbery reviewed (commented):
Looks really good. A bunch of nits and suggestions inline.

ajtowns commented on src/util.cpp:
```diff
@@ -455,6 +455,99 @@ static bool InterpretBool(const std::string& strValue)
+        LOCK(am.cs_args);
+        std::pair<bool,std::string> found_result(false, std::string());
+
+        found_result = GetArgHelper(am.m_override_args, arg, true);
```
I'm not sure; the reason I'm doing it is backwards compatibility. The original reason might have just been a side-effect of using a single variable and updating it every time you see a command line argument, but only if it's not already set when you see a config file argument. I'll add a comment that you can re-nit though. :)

jnewbery replied:
comment seems good, no nits from me :)

ajtowns commented on src/util.cpp:
```diff
@@ -511,55 +658,89 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
-    return strDefault;
+    if (IsArgNegated(strArg)) return "0";
+    std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
+    if (found_res.first) {
```
Yeah, that style is a bit more consistent with other parts of the file.

ajtowns commented on src/util.cpp:
```diff
@@ -748,15 +928,21 @@ void ArgsManager::ReadConfigStream(std::istream& stream)
 {
+    {
+        LOCK(cs_args);
+        m_config_args.clear();
```
I was thinking #10267 would involve an additional call to `ReadConfigStream()` from within `ReadConfigFile()` (ie, you read the config file, find out what additional file(s) to include, then read those, and possibly repeat). I think it makes more sense for `ParseParams()` to only clear the member variable it populates, and `ReadConfigFile()` to clear the member variable it populates. If a later PR wants to change this that's fine, but at least for now it seems fine here to me.

ajtowns commented on src/util.cpp:
```diff
@@ -464,30 +557,84 @@ static bool InterpretBool(const std::string& strValue)
+    } else {
+        ++option_index;
+    }
+    if (key.substr(option_index, 2) == "no") {
```
Fair point, but I'm going to choose to ignore it :)

ajtowns commented:
Fixed a bunch of @jnewbery's nits/complaints/suggestions. Rebased against master to pick up the automated test improvements.

For anyone who got part way through reviewing, piecemeal changes from then to now are on https://github.com/ajtowns/bitcoin/commits/netconf-sections-track with the same underlying tree:

```
$ for a in netconf-sections{,-track}; do git log -1 --pretty="%h -> %T" $a; done
e69767bd55 -> cddf01717c2e081db81a7669decfef7472fe0092
8d892af669 -> cddf01717c2e081db81a7669decfef7472fe0092
```

jnewbery commented:
utACK c25321ff96737bdba80d626d2425ef02c7a4c181. Thanks for addressing all of my nits!

laanwj commented:
I've tested this a bit and it appears to work as expected. I've tested that it's possible to override the port, and proxy configuration per network. I've also tested that putting `testnet=1` in the general section still works.

ryanofsky commented on src/util.cpp:
```diff
@@ -455,6 +455,67 @@ static bool InterpretBool(const std::string& strValue)
+    static inline std::pair<bool,std::string> GetArg(const ArgsManager &am, const std::string& arg)
+    {
+        LOCK(am.cs_args);
+        std::pair<bool,std::string> found_result(false, std::string());
```
boost::optional\<std::string> would be a more natural way to represent this.

ryanofsky commented on src/util.cpp:
```diff
@@ -570,16 +579,19 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
-        m_override_args[key].push_back(val);
+        // Check for -nofoo
+        if (InterpretNegatedOption(key, val)) {
+            m_override_args[key].clear();
```
Shouldn't a negated command line argument clear m_config_args too? As an example, if you had `arg=a` `arg=b` in config the and `-arg=c -noarg -noarg=d`, it would seem weird for the result to be `[a b d]` instead of `[d]`.

ajtowns replied:
ParseParameters is always called prior to ReadConfigFile. Might be worth noting somewhere that you have to do things that way (though it isn't really a choice: if you specify the config file via `-conf`, you don't know what file to pass to ReadConfigFile until you've called ParseParameters).

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11862#discussion_r182556323

> ParseParameters is always called prior to ReadConfigFile

So this means the change I suggested wouldn't actually fix the bug. But you agree this is  introducing a bug that wasn't in @eklitzke's initial implementation from #12713?

ryanofsky replied:
> But you agree this is introducing a bug

Never mind, this bug is actually present in #12713 too because of the initialization order. I guess I thought negated option tracking would provide a clean way to override config file list options instead of just appending to them, but I guess this isn't the way it works.

ryanofsky commented on src/util.cpp:
```diff
@@ -570,16 +579,19 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
+        if (InterpretNegatedOption(key, val)) {
+            m_override_args[key].clear();
+        } else {
+            m_override_args[key].push_back(val);
```
`emplace_back(std::move(val))` would be a little more efficient

ryanofsky commented on src/util.cpp:
```diff
@@ -553,6 +575,11 @@ static bool InterpretNegatedOption(std::string& key, std::string& val)
     return false;
 }
 
+void ArgsManager::SelectConfigNetwork(const std::string& network)
```
Would be slightly more efficiecnt to take a normal (non-reference) string argument, and std::move() it below to avoid a copy.

ryanofsky commented on src/util.cpp:
```diff
@@ -537,6 +537,22 @@ class ArgsManagerHelper {
+                return false; // not set
+            }
+        }
+        return InterpretBool(found_result.second); // is set, so evaluate
```
It seems like code above this line is duplicating code in `ArgsManagerHelper::GetArg` in slightly inconsistent ways (passing getLast `[true, true]` to the two `GetArgHelper` calls instead of `[true, false]` and failing to acquire `cs_args`). Is there a reason for these inconsistencies and for duplicating code here instead of calling getArgs?

ajtowns replied:
Using `ArgsManagerHelper::GetArg` here would mean that calling `GetChainName()` a second (or third or..) time after `SelectParams(GetChainName())` could return a different result than the name of the chain actually being used if people do ridiculous things with their bitcoin.conf.

This should get cleaned up with some of the other config reworking that's going on though.

ajtowns replied:
I think `GetArgHelper(m_config_args, net_arg, true)` probably should have been `, false)`, well spotted.

NicolasDorier commented:
I saw that too late but this PR is a major pain in the ass. Please be more careful with breaking changes.

Not only it broke existing configuration file, but on top of it it call the testnet section `[test]` instead of `[testnet]` .

This mean I can't easily migrate such [templated configuration](https://github.com/btcpayserver/btcpayserver-docker/blob/master/docker-compose-generator/docker-fragments/bitcoin.yml)  without doing major refactoring, or hack in the entrypoint of my docker compose to replace `[test]` by `[testnet]`.

zciendor commented:
Is there even a documentation of the section names? What we have used so far as pseudo-documentation is the config generator from @jlopp https://jlopp.github.io/bitcoin-core-config-generator/
However, it neither contains a `[general]` section like mentioned in https://github.com/bitcoin/bitcoin/pull/11862?_pjax=%23js-repo-pjax-container#issuecomment-381230551 and places the `testnet=1` option currently under the `[debug]` section. This doesn't work anymore and bitcoind seems to completely ignore the `testnet=1` option if it's under the `[debug]` section.

We second @NicolasDorier, 0.17 breaks [our templated approach](https://raw.githubusercontent.com/zhiverbox/armbian-userpatches/master/overlay/opt/zhiverbox/scripts/etc/bitcoin/bitcoin.conf.src) and [install assistant](https://raw.githubusercontent.com/zhiverbox/armbian-userpatches/master/overlay/opt/zhiverbox/scripts/install/50_install-bitcoind.sh) as well, which we use on the [zHIVErbox](https://github.com/zhiverbox/armbian-userpatches/blob/master/README.md) - a low cost, TAILS-inspired security fortess. Looks like a refactoring is needed to support 0.17, which might be OK if it's backwards compatible and the config templates still work on <0.17? But if we need to maintain separate templates for different bitcoin versions it get's ugly.

ryanofsky commented:
Unless I'm mistaken, I think we can restore complete backwards compatibility by just not having "network-only" options, and only warning when one of these options is used outside of a network section, instead of warning and dropping them. I suggested this previously here, https://github.com/bitcoin/bitcoin/issues/14523#issuecomment-431605226.

We could also go further and not even warn about these options if `testnet` is specified in the config file (as opposed to on the command line).

@zciendor, I'm not sure what your comment about a `[debug]` section is referring to. As far as I know that's not a recognized section, so I'd expect any options you enter there to be ignored.

sipa commented:
@ryanofsky I agree, I think we should revert the "ignore certain options outside of network-specific sections" behavior; it seems to gratuitously have broken compatibility.

@zciendor The documentation is in the release notes: https://github.com/bitcoin/bitcoin/blob/v0.17.0/doc/release-notes.md#configuration-sections-for-testnet-and-regtest

jnewbery commented:
I agree that removing the 'network-only' options is a sensible thing to do in order to restore complete backwards compatibility. Network-only options was a good idea to prevent users from accidentally polluting their mainnet config with testnet or regtest config, but the amount of user pain inflicted by not having backwards compatibility means we should revert this change quickly.

To be honest, it didn't even occur to me that this PR would break backwards compatibility. I certainly don't think it was an intentional break.

Alternatively, we could only enforce the 'network-only' rule if the config file contains a network section. That means that old config files would continue to be valid, but new config files with network sections would be protected from certain config errors. That can be added later (after the reversion) if people think that 'network-only' config is a feature worth having.

zciendor commented:
I was talking about the sections the [config file generator](https://jlopp.github.io/bitcoin-core-config-generator/) adds. However, I just noticed that they are commented out, so this should have no effect...

```
# Generated by https://jlopp.github.io/bitcoin-core-config-generator/

# This config should be placed in following path:
# ~/.bitcoin/bitcoin.conf

# [core]
# Keep the transaction memory pool below <n> megabytes.
maxmempool=1500

# [network]
# Use UPnP to map the listening port.
upnp=1

# [zeromq]
# Enable publishing of block hashes to <address>.
zmqpubhashblock=127.0.0.1

# [debug]
# Log IP Addresses in debug output.
logips=1
# Run this node on the Bitcoin Test Network.
testnet=1

# [relay]
# Force relay of transactions from whitelisted peers even if they violate local relay policy.
whitelistforcerelay=0

# [mining]
# Set lowest fee rate (in BTC/kB) for transactions to be included in block creation.
blockmintxfee=0.0001

# [rpc]
# Accept public REST requests.
rest=1

# [wallet]
# Do not load the wallet and disable wallet RPC calls.
disablewallet=1
```

jnewbery commented:
> Is there even a documentation of the section names?

Yes: https://bitcoincore.org/en/releases/0.17.0/#configuration-sections-for-testnet-and-regtest

zciendor commented:
I just realized we seem to have a completely different problem with 0.17. No matter where I specify `testnet=1` (either in the conf file or on the command line) bitcoind tries to write to the default mainnet folder (~/.bitcoin/):

```
$ /usr/local/bin/bitcoind -conf=/etc/bitcoin/test_bitcoin.conf -testnet
Bitcoin Core version v0.17.0 (release build)
InitParameterInteraction: parameter interaction: -proxy set -> setting -discover=0
InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
 
************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       



************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       

Shutdown: In progress...
Shutdown: done

```

The issue is that 0.16 and earlier tried to read/write the /mnt/data/bitcoind/.bitcoin/**testnet3** folder when the `-testnet` (or `testnet=1`) parameter was specified. This seems to have changed and now you have to specify the datadir explicitely along with `testnet=1`?

NicolasDorier commented:
@zciendor unsure if it is related to your problem, but a non obvious consequence of this PR is that on testnet or regtest:

* You can't use Bitcoin-CLI 0.17 on Bitcoind < 0.17
* You can't use Bitcoin-CLI < 0.17 on Bitcoind 0.17

Because they can't read each other config files. I banged my head against the wall for some hour because of this.

ajtowns commented:
@NicolasDorier you can use:

```
    connect=0
    [test]
    connect=0
    [regtest]
    connect=0
```

to have the `connect=0` option (or any of the other mainnet-only options) seen on mainnet/testnet/regtest for pre and post 0.17.

NicolasDorier commented:
I don't think it works, my memory tells that I had issue because 0.16.3 can't parse sections, and thus consider the configuration file invalid.

ajtowns commented:
@NicolasDorier 0.16.3 works fine here with the above -- the boost config parser has always parsed sections, it's just ignored what was in them (because it would treat it as an option named "regtest.connect" which was then never actually used).

jkensik commented:
I get an error about my bitcoin.conf file when I set the port setting in the regtest section. `Error: Config setting for -port only applied on regtest network when in [regtest] section.` 
 This is my bitcoin.conf file, 
`
regtest=1
server=1

[regtest]
port=18333
`

laanwj commented:
@jkensik you are responding to a closed issue from 2018, if you have a problem please report a new issue
